prov/efa/src/rxr/rxr_msg.c

```diff
@@ -93,10 +98,37 @@ ssize_t rxr_msg_post_rtm(struct rxr_ep *rxr_ep, struct rxr_tx_entry *tx_entry)
 		return rxr_pkt_post_ctrl_or_queue(rxr_ep, RXR_TX_ENTRY, tx_entry, rtm_type + tagged, 0);
 	}
 
+	if (delivery_complete_requested_flag && !(peer->flags & RXR_PEER_HANDSHAKE_RECEIVED)) {
+		/*
+		 * The sender cannot send with FI_DELIVERY_COMPLETE
+		 * if the peer is not able to handle it.
+		 * If the sender does not know whether the peer
+		 * can handle it, it needs to wait for
+		 * a handshake packet from the peer. The handshake
+		 * packet contains the information whether the peer
+		 * support it or not.
+		 */
+		if (!(peer->flags & RXR_PEER_HANDSHAKE_RECEIVED)) {
+			err = rxr_pkt_wait_handshake(rxr_ep, tx_entry, peer);
+			if (OFI_UNLIKELY(err))
+				return err;
+		}
+
+		if (delivery_complete_requested_flag &&
+		    peer->flags & RXR_PEER_HANDSHAKE_RECEIVED &&
+		    !rxr_peer_support_delivery_complete(peer) &&
+		    tx_entry->cq_entry.flags & FI_SELECTIVE_COMPLETION)
+			return -FI_EOPNOTSUPP;
+	}
+
 	/* inter instance message */
-	if (tx_entry->total_len <= max_rtm_data_size)
+	if (tx_entry->total_len <= max_rtm_data_size) {
+		ctrl_type = (delivery_complete_requested_flag && rxr_peer_support_delivery_complete(peer)) ?
```

Comment:
Because we already return error for peer not supporting delivery complete, we do not need to check ```rxr_peer_support_delivery_complete(peer)``` again.